### PR TITLE
Special array representation

### PR DESCRIPTION
**Pull request by LakeBlair**
# Overview
Added special array representation for np.zeros and np.identity.

# Details
image

# References
#83 

# Blocked by

### CODE REVIEW

File: examples/examples.ipynb [hunk 1]
```diff
@@ -2,41 +2,20 @@
  "cells": [
```
Comment (odashi) [inline examples/examples.ipynb:2]: (action not required) I guess this notebook file is no longer necessary as we could provide a comprehensive examples in Google Colab. I will remove this file later.

File: src/latexify/codegen/function_codegen.py [hunk 2]
```diff
@@ -356,6 +356,17 @@ def visit_Call(self, node: ast.Call) -> str:
             (default_func_str + r"\mathopen{}\left(", r"\mathclose{}\right)"),
         )
 
+        if func_name == "zeros":
```
Comment (odashi) [inline src/latexify/codegen/function_codegen.py:359]: - For `zeros`, We don't need to constrain only a particular number of dimensions.
- I think these processes don't work if the subtree has unexpected syntax. It usually happens when users gave other functions with the same name. As the AST varies, we basically need complete check of the underlying structure of the given subtree.

**Review by odashi — COMMENTED** [review]
Thanks! It looks this change needs some refactoring to support the feature and error handling completely. Which would you like either:

- fix them by yourself (in this case I will put comprehensive comments)
- delegate remaining tasks to me (your contribution is still recorded correctly)

**Comment by odashi** [comment]
@LakeBlair Let's continue developing this pull request as you preferred in another thread.

Before providing additional comments, please resolve the following issues:

- Merge main into this branch and resolve all conflicts.
- run `./check.sh` and resolve all errors.

**Comment by LakeBlair** [comment]
> @LakeBlair Let's continue developing this pull request as you preferred in another thread.
> 
> Before providing additional comments, please resolve the following issues:
> 
> * Merge main into this branch and resolve all conflicts.
> * run `./check.sh` and resolve all errors.

Just did both. Check if they look good to you.

File: src/latexify/codegen/function_codegen.py [hunk 3]
```diff
@@ -425,7 +425,26 @@ def visit_Call(self, node: ast.Call) -> str:
...
+            str = ""
+            open_bracket = "{"
+            close_bracket = "}"
+            for i, elt in enumerate(node.args[0].elts):
+                str += "{" + self.visit(elt) + "}"
+                if i != len(node.args[0].elts) - 1:
+                    str += " " + r"\times" + " "
+
+            matrix_str = "0^" + open_bracket + str + close_bracket
+            return matrix_str
```
Comment (odashi) [inline src/latexify/codegen/function_codegen.py:440]: General suggestions:

- Requires complete syntax checking.
- Don't use `str` as a variable as it is reserved by the builtin type name. Overwriting builtins will confuse the behavior of the code (even if the current code doesn't involve any errors, it will happen in the future). Use `latex` instead if you need to store some generated strings.
- Suppress string concatenation as [it is significantly expensive](https://stackoverflow.com/a/3055541). Making a sequence (generator), then join all of them is generally better.
- I think `\mathbf{0}` ( $\mathbf{0}$ ) should be used to distinguish that this is not a scalar $0$.

```suggestion
            if len(node.args) != 1:
                # fall back to the default.
                
            arg0 = node.args[0]
                
            if not isinstance(arg0, ast.Tuple):
                # fall back to the default.
                # Tecunically we don't support `zeros(n)` where `n` is a scalar.
            
            dims_latex = r" \times ".join(self.visit(x) for x in arg0.elts)
            return fr"\mathbf{{0}}^{{{dims_latex}}}"
```

File: src/latexify/codegen/function_codegen_test.py [hunk 4]
```diff
@@ -862,6 +862,34 @@ def test_use_set_symbols_compare(code: str, latex: str) -> None:
...
+    tree = ast.parse(
+        textwrap.dedent(
+            """
+        def f(a, b):
+            return np.zeros((a,b))
+            """
+        )
+    ).body[0]
+    latex = "f(a, b) = 0^{{a} \\times {b}}"
+    assert isinstance(tree, ast.FunctionDef)
```
Comment (odashi) [inline src/latexify/codegen/function_codegen_test.py:875]: We don't need complete parsing. It is enough to obtain only `ast.Call`.
```suggestion
    tree = ast_utils.parse_expr("zeros((a, b))")
    assert isinstance(tree, ast.Call)
```

File: src/latexify/codegen/function_codegen_test.py [hunk 5]
```diff
@@ -862,6 +862,34 @@ def test_use_set_symbols_compare(code: str, latex: str) -> None:
     assert function_codegen.FunctionCodegen(use_set_symbols=True).visit(tree) == latex
 
 
+def test_generate_numpy_zeros():
```
Comment (odashi) [inline src/latexify/codegen/function_codegen_test.py:865]: Please write comprehensive tests to cover every edge cases, as in other parametric tests.

File: src/latexify/codegen/function_codegen_test.py [hunk 6]
```diff
@@ -862,6 +862,34 @@ def test_use_set_symbols_compare(code: str, latex: str) -> None:
...
+            return np.zeros((a,b))
+            """
+        )
+    ).body[0]
+    latex = "f(a, b) = 0^{{a} \\times {b}}"
+    assert isinstance(tree, ast.FunctionDef)
+    assert FunctionCodegen().visit(tree) == latex
+
+
+def test_generate_numpy_identity():
```
Comment (odashi) [inline src/latexify/codegen/function_codegen_test.py:879]: ditto

File: src/latexify/codegen/function_codegen.py (hunk 3 above)
Comment (LakeBlair) [reply src/latexify/codegen/function_codegen.py]: @odashi I made some changes to my code. 

- np.zeros and np.identity are now in a new function, treated as special numpy methods.
- I added some if statements checking for valid ast inputs. 
- I added some parametric tests. 

Some more considerations:

- Some string concatenation is simplified, some could be simplified further.
- In the case of np.zeros(0), I just represent it as $\mathbf{0}^{{1} \times {0}}$. There might be better ways of expressing it as a special case. Do you have any suggestions?

Let me know what you think.

File: src/latexify/codegen/function_codegen.py [hunk 7]
```diff
@@ -416,6 +416,49 @@ def generate_matrix_from_array(data: list[list[str]]) -> str:
 
         return generate_matrix_from_array(rows)
 
+    def _generate_special_latex(self, node: ast.Call, func_name) -> str | None:
```
Comment (odashi) [inline src/latexify/codegen/function_codegen.py:419]:
```suggestion
    def _generate_special_latex(self, node: ast.Call, func_name: str) -> str | None:
```

**Review by odashi — COMMENTED** [review]
Sorry I still think this change requires several refactoring, I will do it by my side.

**Review by odashi — APPROVED** [review]
